### DISCUSSION TITLE
Lifespan Events Not Triggering on Server Deployment, but Functioning Locally

### DISCUSSION THREAD

**Discussion by Abhishekkumar021** [fastapi/fastapi#13008] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app/main.py

from fastapi import FastAPI
from contextlib import asynccontextmanager
from fastapi.middleware.cors import CORSMiddleware
from dotenv import load_dotenv
import os
from app.db.db_connection import connect_to_mongo, close_mongo_connection
from app.api.user_routes import user_router
from app.api.product_routes import product_router
from fastapi.staticfiles import StaticFiles

load_dotenv()


@asynccontextmanager
async def db_lifespan(app: FastAPI):
    # Startup

    database = await connect_to_mongo()
    ping_response = await database.command("ping")
    if int(ping_response["ok"]) != 1:
        raise Exception("Problem connecting to database cluster.")
    else:
        print("Connected to database cluster.")

    # shutdown
    yield
    await close_mongo_connection()


app: FastAPI = FastAPI(lifespan=db_lifespan)

app.add_middleware(
    CORSMiddleware,
    allow_origins=os.getenv("Origin", "*"),
    # allow_origins=origins,
    allow_credentials=True,
    allow_methods=os.getenv("Allow_Methods", "*"),
    # allow_methods=["GET", "POST", "PUT", "DELETE", "OPTIONS"],
    allow_headers=["*"],
)

static_dir = os.path.join(os.getcwd(), "static")

# Mount the static directory at '/static'
app.mount("/static", StaticFiles(directory=static_dir), name="static")


@app.get("/")
def home() -> dict:
    return {"Hello": "World"}


app.include_router(user_router, prefix="/api/v1/user")

app.include_router(product_router, prefix="/api/v1/products")

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)



app/db/db_connection.py

from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase
from pymongo.errors import ConnectionFailure
from typing import Any
from dotenv import load_dotenv
import os
from urllib.parse import quote_plus

load_dotenv()


class MongoDB:
    client: AsyncIOMotorClient | None = None
    database: AsyncIOMotorDatabase | Any = Any


db = MongoDB()


async def connect_to_mongo() -> AsyncIOMotorDatabase[Any]:

    try:

        username = quote_plus(str(os.getenv("MONGODB_USERNAME")))
        password = quote_plus(str(os.getenv("MONGODB_PASSWORD")))
        mongoDB_cluster = quote_plus(str(os.getenv("MONGODB_CLUSTER")))
        mongoDB_auth_mechanism = quote_plus(str(os.getenv("MONGODB_AUTH_MACHANISM")))

        URI = f"mongodb+srv://{username}:{password}@{mongoDB_cluster}.oufdzrs.mongodb.net/?authMechanism={mongoDB_auth_mechanism}"
        db.client = AsyncIOMotorClient(URI)
        db.database = db.client[str(os.getenv("DB_NAME"))]
        print("MongoDB connected!")

        return db.database

    except ConnectionFailure as e:

        print(f"Failed to connect to MongoDB: {e}")
        raise Exception("Problem connecting to database")


async def close_mongo_connection():
    if db.client:
        db.client.close()
        print("MongoDB connection closed!")


#other routes inside

app/api/(.py file...).
```

### Description

I am currently facing an issue with my **FastAPI** application where the **lifespan events** (startup and shutdown) are functioning correctly on my local machine but are not triggering as expected when I deploy the application to a server (specifically on Vercel).

When I run the application locally, I can see the expected print statements in the console, indicating that the connection to **MongoDB** is established during startup and closed during shutdown.

However, upon **deploying the same application to Vercel,** I do not see these print statements in the logs, and it appears that the MongoDB connection is not being established. I have verified that all necessary **environment** **variables** are correctly set in the Vercel dashboard.

Furthermore, I am experiencing an issue where the MongoDB connection is not available in my other route files. While it works when I explicitly call` connect_to_mongo()` in my **home route** `@app.get("/")` , it does not work in other routes unless I call it again. I want to ensure that the database connection is shared across all routes without needing to call the connection function in each one.

I appreciate any insights or suggestions you may have. Thank you for your help!

@tiangolo @

### Operating System

Windows

### FastAPI Version

0.114.0

### Pydantic Version

2.9.1

### Python Version

Python 3.12.4

### Additional Context

![Screenshot 2024-11-28 190556](https://github.com/user-attachments/assets/048d2f6c-4b7a-4908-a620-9e349a922dbf)

I attached console snapshot for reference

**Comment by nag763** [comment]
Hey @Abhishekkumar021 
Actually did have a look on my end this morning as I met the same issue,
Note that my knowledge of how FastAPI and its associated runtimes is pretty limited.

Checking the vercel documentation there on python runtime, especially at the section of the link [here](https://vercel.com/docs/functions/runtimes/python#advanced-python-usage), I note that it says there are two ways for vercel to run serverless :

> The entry point of this runtime is a glob matching .py source files with one of the following variables defined:
>    * `handler` that inherits from the BaseHTTPRequestHandler class
 >   * `app` that exposes a WSGI or ASGI Application

Basicly I can see the same behavior happening while using gunicorn over uvicorn (sorry no sample code to provide)

And the same behavior can be noticed while starting development with `vercel dev` command.

My assumption is that vercel is using a runtime that isn't reading properly the event as it is outside the app context and its route definitions.

The only workaround I have is to have the event outside any FastAPI application context, which seem to be working better (myself I was attempting in the context of alembic migrations) :

```diff
- @asynccontextmanager
- async def lifespan(app: FastAPI):
    alembic.config.main(["upgrade", "head"])
    logger.info("Alembic migration ran with success")
-  yield
```

If anyone knows how to request Vercel team directly to have a look on this issue, gently let me know!

**Comment by yakubka** [comment]
Lifespan events not triggering on a server deployment but working locally almost always means the ASGI server is not calling the lifespan protocol on your application. Some deployment setups (nginx unit, some Gunicorn configurations, serverless platforms) run the application without sending the lifespan startup scope.

Uvicorn sends the lifespan scope by default. Gunicorn with `uvicorn.workers.UvicornWorker` also sends it. The issue usually appears when a reverse proxy or platform intercepts the process startup without going through the ASGI lifespan protocol.

To verify, add a log line at the top of your lifespan function.

```python
import logging
from contextlib import asynccontextmanager

@asynccontextmanager
async def lifespan(app):
    logging.warning("LIFESPAN STARTUP")
    yield
    logging.warning("LIFESPAN SHUTDOWN")
```

If you do not see the log on the server, the lifespan is not being called. The fix depends on your deployment. For Gunicorn, make sure you use `UvicornWorker`. For serverless platforms that do not support lifespan, move your startup logic to a module-level call that runs when the module is imported rather than inside the lifespan function.
